enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,300 @@
+As a mitigation, Device Administrators should reinstall the upgrade when network
+is available for the Edge Devices.
+
+To further mitigate potential problems related to slow or unstable networks,
+Device Administrators may choose to bundle the upgraded container images in the
+`ostree` software upgrade. In this case, when the upgrade is installed, no
+more container images will need to be downloaded over the network.
+
+Some 3rd party workloads may become operational before the upgrade is declared
+valid and potentially create or update data on the device. If a rollback is
+performed subsequentially, there is a risk of data loss because the file system
+is reverted to its state before the upgrade. One of the ways to mitigate this 
+problem is to have 3rd party workloads wait until a boot is declared successful.
+
+```bash
+$ sudo grub2-editenv - list | grep ^boot_success
+boot_success=1
+```
+
+If any of the user workloads start before the boot is declared successful and
+update host file system or data volumes, there is a risk for the host file system
+and software update to be reverted to its previous state by `greenboot`. In this
+case, user workloads should take measures to ensure their configuration and
+application data consistency. For example, user application may choose to install
+scripts in the `/etc/greenboot/red.d` directory to be invoked by `greenboot` in
+case of an unsuccessful upgrade before the rollback is performed.
+
+> MicroShift implements its own application state and data backup / rollback
+> procedures, but it cannot ensure consistency of 3rd party workloads or data.
+
+### Drawbacks
+
+The technical implementation of upgrade validation is rather simple as it relies
+on the existing `greenboot` and `rpm-ostree` facilities. However, end users may
+find it annoying to have their Edge Devices rebooted few times until the upgrade
+is either successful or rolled back.
+
+It is possible that certain 3rd party workloads become operational before the
+upgrade is declared successful. This situation may further hurt user experience
+if their Edge Devices are rebooted due to MicroShift core functionality
+validation not passing.
+
+## Design Details
+
+### Open Questions
+N/A
+
+### Test Plan
+N/A
+
+### Graduation Criteria
+N/A
+
+#### Dev Preview -> Tech Preview
+N/A
+
+#### Tech Preview -> GA
+N/A
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+
+The validation scripts will be delivered as part of the MicroShift core RPM
+packages. Any problems caused by the ugraded software, scripts or system
+configuration will be rolled back by `greenboot` when validations do not pass.
+
+> The mutable parts of the file system and data volumes are not rolled back
+> automatically by the `ostree`. Another enhancement should deal with that issue.
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A
+
+#### Failure Modes
+N/A
+
+#### Support Procedures
+N/A
```

Comment:
We can refer to the specific logs we would recommend examining, right?

Reply:
Correct. Added.